src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -612,6 +671,157 @@ private BoundExpression CreateAndPopulateArray(BoundCollectionExpression node, A
                 arrayType);
         }
 
+        /// <summary>
+        /// Returns true if type is convertible to Span or ReadOnlySpan.
+        /// If non-identity conversion, also returns a non-null asSpanMethod.
+        /// </summary>
+        private bool TryGetSpanConversion(TypeSymbol type, out MethodSymbol? asSpanMethod)
+        {
+            if (type is NamedTypeSymbol spanType
+                && (spanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Span_T), TypeCompareKind.ConsiderEverything)
+                    || spanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.ConsiderEverything)))
+            {
+                asSpanMethod = null;
+                return true;
+            }
+
+            if (type is ArrayTypeSymbol { IsSZArray: true } arrayType
+                && _factory.WellKnownMethod(WellKnownMember.System_Span_T__ctor_Array, isOptional: true) is { } spanCtorArray)
+            {
+                asSpanMethod = spanCtorArray.AsMember(spanCtorArray.ContainingType.Construct(arrayType.ElementType));
+                return true;
+            }
+
+            if (type is NamedTypeSymbol immutableArrayType
+                && immutableArrayType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Collections_Immutable_ImmutableArray_T), TypeCompareKind.ConsiderEverything)
+                && _factory.WellKnownMethod(WellKnownMember.System_Collections_Immutable_ImmutableArray_T__AsSpan, isOptional: true) is { } immutableArrayAsSpanMethod)
+            {
+                asSpanMethod = immutableArrayAsSpanMethod!.AsMember(immutableArrayType);
+                return true;
+            }
+
+            if (type is NamedTypeSymbol listType
+                && listType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_List_T), TypeCompareKind.ConsiderEverything)
+                && _factory.WellKnownMethod(WellKnownMember.System_Runtime_InteropServices_CollectionsMarshal__AsSpan_T, isOptional: true) is { } collectionsMarshalAsSpanMethod)
+            {
+                asSpanMethod = collectionsMarshalAsSpanMethod.Construct(listType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type);
+                return true;
+            }
+
+            asSpanMethod = null;
+            return false;
+        }
+
+        private BoundExpression? TryConvertToSpanOrReadOnlySpan(BoundExpression expression)
+        {
+            var type = expression.Type;
+            Debug.Assert(type is not null);
+
+            if (!TryGetSpanConversion(type, out var asSpanMethod))
+            {
+                return null;
+            }
+
+            return asSpanMethod is null ? expression : CallAsSpanMethod(expression, asSpanMethod);
```

Comment:
>asSpanMethod is null ? expression : CallAsSpanMethod(expression, asSpanMethod) [](http://example.com/codeflow?start=19&length=78)

It looks like all calls to `CallAsSpanMethod()` use this same pattern. Perhaps rename the method to `CallOptionalAsSpanMethod(expression, asSpanMethod)` and allow `asSpanMethod` to be `null`, to simplify callers.